File: VRSampleApp/VRSampleApp/Classes/PresentationLayer/Modules/ListOfVideosModule/View/Controller/ListOfVideosViewController.m
//
//  ListOfVideosViewController.m
//  VRSampleApp
//
//  Copyright © 2017 Ooyala Inc. All rights reserved.
//

#import "ListOfVideosViewController.h"


@interface ListOfVideosViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic) UISwitch *QAModeSwitch;

@end


@implementation ListOfVideosViewController

#pragma mark - Life cycle

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self configureNavigationBar];
}

#pragma mark - Private functions

- (void)configureNavigationBar {
  
  // Add QA switch
  
  self.QAModeSwitch = [UISwitch new];
  UILabel *QALabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 44, 44)];
  
  QALabel.textAlignment = NSTextAlignmentRight;
  QALabel.text = @"QA";
  
  UIBarButtonItem *QASwitchBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.QAModeSwitch];
  UIBarButtonItem *QALabelBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:QALabel];
  
  // Add custom video bar button item
  
  UIBarButtonItem *customVideoBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                                                                            target:self
                                                                                            action:@selector(customVideoBarButtonAction:)];
  
  // Add elements to navigation bar
  
  self.navigationItem.rightBarButtonItems = @[QASwitchBarButtonItem, QALabelBarButtonItem];
  self.navigationItem.leftBarButtonItems = @[customVideoBarButtonItem];
}

- (void)showCustomVideoItemViewController {
  __typeof(self) __weak weakSelf = self;
  
  UIViewController *customVideoViewController = [self.viewModel configuredCustomVideoViewControllerWithCompletion:^(VideoItem *videoItem) {
    [weakSelf openVideoPlayerModuleWithVideoItem:videoItem];
  }];
  
  [self presentViewController:customVideoViewController animated:YES completion:NULL];
}

- (void)openVideoPlayerModuleWithVideoItem:(VideoItem *)videoItem {
  UIViewController *viewController = [self.viewModel configuredVideoViewControllerWithVideoItem:videoItem
                                                                               andQAModeEnabled:self.QAModeSwitch.isOn];
  
  [self.navigationController pushViewController:viewController animated:YES];
}

#pragma mark - Actions

- (void)customVideoBarButtonAction:(id)sender {
  [self showCustomVideoItemViewController];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return [self.viewModel getCountSections];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return [self.viewModel getCountRowsInSection:section];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ListOfVideosViewControllerTableViewCellID"];
  VideoItem *videoItem = [self.viewModel getVideoItemAt:indexPath];
  
  if (!cell) {
    cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"ListOfVideosViewControllerTableViewCellID"];
  }
  
  if (videoItem) {
    cell.textLabel.text = videoItem.title;
  } else {
    cell.textLabel.text = @"Unknow video";
  }
  
  return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
  VideoItemSection *videoItemSection = [self.viewModel getVideoItemSectionAtSection:section];
  
  if (videoItemSection) {
    return videoItemSection.title;
  }
  
  return @"Unknow section title";
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  VideoItem *videoItem = [self.viewModel getVideoItemAt:indexPath];
  
  if (videoItem) {
    [self openVideoPlayerModuleWithVideoItem:videoItem];
  }
}


@end

File: SwiftVRSkinSampleApp/SwiftVRSkinSampleApp/Classes/PresentationLayer/Modules/ListOfVideosModule/View/ViewController/ListOfVideosViewController.swift
//
//  ListOfVideosViewController.swift
//  VRSkinSampleApp
//
//  Copyright © 2017 Ooyala, Inc. All rights reserved.
//

import UIKit


class ListOfVideosViewController: UIViewController {
  
  // MARK: - Public properties
  
  var viewModel: ListOfVideosViewModel!
  
  // MARK: - Private properties
  private var QAModeSwitch: UISwitch!
  
  // MARK: - Life cycle
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    configureNavigationBar()
  }
  
  // MARK: - Private functions
  
  private func configureNavigationBar() {
    
    // Add QA switch
    
    QAModeSwitch = UISwitch()
    let QALabel = UILabel(frame: CGRect(x: 0, y: 0, width: 44, height: 44))
    
    QALabel.textAlignment = .right
    QALabel.text = "QA"
    
    let QASwitchBarButtonItem = UIBarButtonItem(customView: QAModeSwitch)
    let QALabelBarButtonItem = UIBarButtonItem(customView: QALabel)
    
    // Add custom video bar button item
    
    let customVideoBarButtonItem = UIBarButtonItem(
      barButtonSystemItem: .add, target: self, action: #selector(customVideoBarButtonAction))
    
    // Add elements to navigation bar
    
    navigationItem.rightBarButtonItems = [QASwitchBarButtonItem, QALabelBarButtonItem]
    navigationItem.leftBarButtonItems = [customVideoBarButtonItem]
  }
  
  private func showCustomVideoItemViewController() {
    let customVideoViewController = viewModel.configuredCustomVideoViewController(completion: { videoItem in
      self.openVideoPlayerModule(withVideoItem: videoItem)
    })
    
    self.present(customVideoViewController, animated: true, completion: nil)
  }
    
  private func openVideoPlayerModule(withVideoItem: VideoItem) {
    let viewController = viewModel.configuredVideoViewController(
      withVideoItem: withVideoItem, QAModeEnabled: QAModeSwitch.isOn)
    
    navigationController?.pushViewController(viewController, animated: true)
  }
  
  // MARK: - Actions
  
  @objc private func customVideoBarButtonAction(_ sender: Any?) {
    showCustomVideoItemViewController()
  }
  
  
}

// MARK: - UITableViewDataSource

extension ListOfVideosViewController: UITableViewDataSource {
  
  func numberOfSections(in tableView: UITableView) -> Int {
    return viewModel.getCountSections()
  }
  
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return viewModel.getCountRowsIn(section: section)
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    var cell = tableView.dequeueReusableCell(withIdentifier: "ListOfVideosViewControllerTableViewCellID")
    
    if cell == nil {
      cell = UITableViewCell(style: .default, reuseIdentifier: "ListOfVideosViewControllerTableViewCellID")
    }
    
    if let videoItem = viewModel.getVideoItemAt(indexPath: indexPath) {
      cell?.textLabel?.text = videoItem.title
    } else {
      cell?.textLabel?.text = "Unknow video"
    }
    
    return cell!
  }
  
  func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
    if let videItemSection = viewModel.getVideoItemSectionAt(section: section) {
      return videItemSection.title
    }
    
    return "Unknow section title"
  }
  
  
}

// MARK: - UITableViewDelegate

extension ListOfVideosViewController: UITableViewDelegate {
  
  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    if let videoItem = viewModel.getVideoItemAt(indexPath: indexPath) {
      openVideoPlayerModule(withVideoItem: videoItem)
    }
  }
  
  
}
